apps/designer/app/routes/rest/pages.$projectId.ts

```diff
@@ -1,6 +1,89 @@
-import type { LoaderFunction } from "@remix-run/node";
+import { LoaderFunction, ActionFunction } from "@remix-run/node";
 import { db } from "@webstudio-is/project/index.server";
-import { Pages } from "@webstudio-is/project";
+import { type Pages, utils } from "@webstudio-is/project";
+import { zfd } from "zod-form-data";
+import { z } from "zod";
+import { CreatePageData } from "~/designer/features/sidebar-left/panels/pages/settings";
+import { sentryException } from "~/shared/sentry";
+
+const nonEmptyString = z
+  .string({
+    invalid_type_error: "Can't be empty",
+    required_error: "Can't be empty",
+  })
+  .transform((v) => v.trim())
+  .refine((val) => val !== "", "Can't be empty");
+
+const CreatePageInput = zfd.formData({
+  name: nonEmptyString,
+  path: nonEmptyString
+    .refine((path) => path !== "/", "Can't be just a /")
+    .transform((path) => (path.endsWith("/") ? path.slice(0, -1) : path))
+    .transform((path) => path.replace(/\/\//g, "/"))
+    .refine(
+      (path) => path === "" || path.startsWith("/"),
+      "Must start with a /"
+    )
+    .refine(
+      (path) => /^[-_a-z0-9\\/]*$/.test(path),
+      "Only a-z, 0-9, -, _ and / are allowed"
+    ),
+});
+
+const handlePUT = async (
+  projectId: string,
+  request: Request
+): Promise<CreatePageData> => {
+  const result = CreatePageInput.safeParse(await request.formData());
+  if (result.success === false) {
+    return { errors: result.error.formErrors };
+  }
+  const data = result.data;
+
+  const devBuild = await db.build.loadByProjectId(projectId, "dev");
+
+  const existingPage = utils.pages.findByPath(devBuild.pages, data.path);
+  if (existingPage !== undefined) {
+    return {
+      errors: {
+        formErrors: [],
+        fieldErrors: {
+          path: [`Already used for "${existingPage.name}"`],
+        },
+      },
+    };
+  }
+
+  const updatedBuild = await db.build.addPage(devBuild.id, data);
+
+  const newPage = utils.pages.findByPath(updatedBuild.pages, data.path);
+
+  if (newPage === undefined) {
+    throw new Error("New page not found");
+  }
+
+  return { ok: true, page: newPage };
+};
+
+export const action: ActionFunction = async ({ request, params }) => {
+  try {
+    if (params.projectId === undefined) {
+      throw new Error(`Project ID required`);
+    }
+
+    if (request.method.toLowerCase() === "put") {
```

Comment:
isn't it consistently upper case from remix? So it doesn't depend what you used to send it, would be simpler to use it as-is instead of converting the case every time, well unless it's actually not consistent

Reply:
I don't know whether it's consistent. This is not documented as far as I know

Reply:
Actually I just remembered that I was using a lower case POST/DELETE on the client and received upper case on the server, so it is consistent

Reply:
in the sense that on the request.method its always upper case